libs/rays/src/PathTracer.cpp

```diff
@@ -0,0 +1,480 @@
+        context->config.tileCallback = [] (image::LinearImage target,
+                filament::math::ushort2 topLeft, filament::math::ushort2 bottomRight,
+                void* userData) {
+        };
+    }
+
+    if (!context->config.doneCallback) {
+        context->config.doneCallback = [] (image::LinearImage target, void* userData) {
+        };
+    }
+
+    // Create the embree device.
+    RTCDevice device = context->device = rtcNewDevice(nullptr);
+    rtcSetDeviceErrorFunction(device, [](void* userPtr, RTCError code, const char* str) {
+        printf("Embree error: %s.\n", str);
+    }, nullptr);
+
+    // Populates an embree scene from 3D position data, ignoring normals and UVs.
+    auto populate3DScene = [=] () {
+        for (size_t i = 0; i < context->config.numMeshes; ++i) {
+            const SimpleMesh& mesh = context->config.meshes[i];
+            RTCGeometry geo = rtcNewGeometry(device, RTC_GEOMETRY_TYPE_TRIANGLE);
+            rtcSetSharedGeometryBuffer(geo, RTC_BUFFER_TYPE_VERTEX, 0, RTC_FORMAT_FLOAT3,
+                    mesh.positions, 0, mesh.positionsStride, mesh.numVertices);
+            rtcSetSharedGeometryBuffer(geo, RTC_BUFFER_TYPE_INDEX, 0, RTC_FORMAT_UINT3,
+                    mesh.indices, 0, sizeof(uint32_t) * 3, mesh.numIndices / 3);
+            rtcCommitGeometry(geo);
+            rtcAttachGeometry(context->scene, geo);
+            rtcReleaseGeometry(geo);
+        }
+        rtcCommitScene(context->scene);
+    };
+
+    // Populates an embree scene from 2D position data sourced from UV rather than the standard 3D
+    // vertex positions.
+    auto populate2DScene = [=] () {
+        for (size_t i = 0; i < context->config.numMeshes; ++i) {
+            const SimpleMesh& mesh = context->config.meshes[i];
+            RTCGeometry geo = rtcNewGeometry(device, RTC_GEOMETRY_TYPE_TRIANGLE);
+            rtcSetGeometryVertexAttributeCount(geo, 2);
+            rtcSetSharedGeometryBuffer(geo, RTC_BUFFER_TYPE_VERTEX, 0, RTC_FORMAT_FLOAT3,
+                    mesh.uvs, 0, mesh.uvsStride, mesh.numVertices);
+            rtcSetSharedGeometryBuffer(geo, RTC_BUFFER_TYPE_VERTEX_ATTRIBUTE, 0, RTC_FORMAT_FLOAT3,
+                    mesh.positions, 0, mesh.positionsStride, mesh.numVertices);
+            rtcSetSharedGeometryBuffer(geo, RTC_BUFFER_TYPE_VERTEX_ATTRIBUTE, 1, RTC_FORMAT_FLOAT3,
+                    mesh.normals, 0, mesh.normalsStride, mesh.numVertices);
+            rtcSetSharedGeometryBuffer(geo, RTC_BUFFER_TYPE_INDEX, 0, RTC_FORMAT_UINT3,
+                    mesh.indices, 0, sizeof(uint32_t) * 3, mesh.numIndices / 3);
+            rtcCommitGeometry(geo);
+            rtcAttachGeometry(context->scene, geo);
+            rtcReleaseGeometry(geo);
+        }
+        rtcCommitScene(context->scene);
+    };
+
+    if (mConfig.uvCamera) {
+
+        context->scene = rtcNewScene(device);
+        populate2DScene();
+
+        context->gbuffer = image::LinearImage(width, height, 6);
```

Comment:
Maybe a comment explaining the layout of the `gbuffer`? Is it 3 position floats followed by 3 normal floats?

Reply:
Done.